src/Core/Domain/ValueObject/Reduction.php

```diff
@@ -0,0 +1,160 @@
+     */
+    const TYPE_PERCENTAGE = 'percentage';
+
+    /**
+     * Allowed reduction types
+     */
+    const ALLOWED_TYPES = [
+        self::TYPE_AMOUNT,
+        self::TYPE_PERCENTAGE,
+    ];
+
+    /**
+     * Maximum allowed value for percentage type reduction
+     */
+    const MAX_ALLOWED_PERCENTAGE = 100;
+
+    /**
+     * @var string
+     */
+    private $type;
+
+    /**
+     * @var Number
+     */
+    private $value;
+
+    /**
+     * @param string $type
+     * @param float $value
+     *
+     * @throws DomainConstraintException
+     */
+    public function __construct(string $type, float $value)
+    {
+        $this->assertIsAllowedType($type);
+        $this->assertIsValidValue($type, $value);
+        $this->type = $type;
+        $this->value = new Number((string) $value);
+    }
+
+    /**
+     * @return string
+     */
+    public function getType(): string
+    {
+        return $this->type;
+    }
+
+    /**
+     * @return Number
+     */
+    public function getValue(): Number
+    {
+        return $this->value;
+    }
+
+    /**
+     * @param string $type
+     *
+     * @throws DomainConstraintException
+     */
+    private function assertIsAllowedType(string $type)
+    {
+        if (!in_array($type, self::ALLOWED_TYPES, true)) {
+            throw new DomainConstraintException(
+                sprintf(
+                    'The reduction type "%s" is invalid. Valid types are: "%s", "%s".',
+                    $type,
+                    self::TYPE_AMOUNT,
+                    self::TYPE_PERCENTAGE
+                ),
+                DomainConstraintException::INVALID_REDUCTION_TYPE
+            );
+        }
+    }
+
+    /**
+     * @param $type
+     * @param $value
+     *
+     * @throws DomainConstraintException
+     */
+    private function assertIsValidValue(string $type, float $value)
+    {
+        if (self::TYPE_PERCENTAGE === $type) {
+            if (!$this->assertIsNotNegative($value) || self::MAX_ALLOWED_PERCENTAGE < $value) {
+                throw new DomainConstraintException(
+                    sprintf(
+                        'Invalid reduction percentage "%s". It must be from 0 to %s%%',
+                        $value,
+                        self::MAX_ALLOWED_PERCENTAGE),
+                    DomainConstraintException::INVALID_REDUCTION_PERCENTAGE
+                );
+            }
+        }
+
+        if (!$this->assertIsNotNegative($value)) {
+            throw new DomainConstraintException(
+                sprintf(
+                    'Invalid reduction amount "%s". It cannot be less than 0',
+                    $value
+                ),
+                DomainConstraintException::INVALID_REDUCTION_AMOUNT
+            );
+        }
+    }
+
+    /**
+     * @param float $value
+     *
+     * @return bool
+     */
+    private function assertIsNotNegative(float $value)
+    {
+        return 0 <= $value;
```

Comment:
I might be paranoid, but should'nt it be `0.0 <= value` ? when $value is compared to int `0` I'm afraid there might be a casting to int so `0.5` would be considered a negative value

Reply:
just tested it, result is as expected. 0 <= 0.5 is true.